src/gwt/src/org/rstudio/studio/client/workbench/views/source/Source.java

```diff
@@ -538,6 +544,88 @@ public void onExecuteUserCommand(ExecuteUserCommandEvent event)
       events.addHandler(DocWindowChangedEvent.TYPE, this);
       events.addHandler(DocTabDragInitiatedEvent.TYPE, this);
       events.addHandler(PopoutDocInitiatedEvent.TYPE, this);
+      
+      events.addHandler(
+            ReplaceSelectionEvent.TYPE,
+            new ReplaceSelectionEvent.Handler()
+            {
+               @Override
+               public void onReplaceSelection(ReplaceSelectionEvent event)
+               {
+                  if (activeEditor_ == null)
+                     return;
+                  
+                  if (!(activeEditor_ instanceof TextEditingTarget))
+                     return;
+                  
+                  TextEditingTarget target = (TextEditingTarget) activeEditor_;
+                  target.insertCode(event.getData(), false);
+               }
+            });
+      
+      events.addHandler(
+            ReplaceRangesEvent.TYPE,
+            new ReplaceRangesEvent.Handler()
+            {
+               @Override
+               public void onReplaceRanges(ReplaceRangesEvent event)
+               {
+                  if (activeEditor_ == null)
+                     return;
+                  
+                  if (!(activeEditor_ instanceof TextEditingTarget))
+                     return;
+                  
+                  TextEditingTarget target = (TextEditingTarget) activeEditor_;
+                  
+                  JsArray<ReplacementData> data = event.getData().getReplacementData();
+                  for (int i = 0; i < data.length(); i++)
+                  {
+                     ReplacementData el = data.get(i);
+                     Range range = el.getRange();
+                     String text = el.getText();
+                     
+                     target.getDocDisplay().replaceRange(range, text);
+                  }
+               }
+            });
+      
+      events.addHandler(
+            GetActiveDocumentContextEvent.TYPE,
```

Comment:
Without some modification I think this is going to cause all source windows to report their active context, which is probably not what you want. I'd put this event handler in the source window manager and have it dispatch to the source shim for the active window. (Alternately, you could say that this method only works on the active doc in the main window, in which case you could just ignore this in source windows.)